dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +232,412 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
             }
 
-            // Check for COM flag, a valid cookie
-            if (WI_IsFlagSet(activators, PushNotificationRegistrationActivators::ComActivator))
+            auto scope_exit = wil::scope_exit(
+                [&] {
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_registering = false;
+                });
+
+            if (AppModel::Identity::IsPackagedProcess())
             {
-                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !s_comActivatorRegistration, "ComActivator not registered.");
-                s_comActivatorRegistration.reset();
+                auto lock{ m_lock.lock_exclusive() };
+                THROW_HR_IF_MSG(E_UNEXPECTED, !m_comActivatorRegistration, "Need to call Register() before calling Unregister().");
+                m_comActivatorRegistration.reset();
             }
 
-            if (WI_IsFlagSet(activators, PushNotificationRegistrationActivators::ProtocolActivator))
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
-                THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !s_protocolRegistration);
-                auto coInitialize = wil::CoInitializeEx();
+                auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+                THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
+            }
+            else
+            {
+                std::wstring processName;
+                {
+                    auto lock{ m_lock.lock_shared() };
+                    THROW_HR_IF_MSG(E_UNEXPECTED, !m_singletonForegroundRegistration, "Need to call Register() before calling Unregister().");
+                    processName = m_processName.get();
+                }
 
-                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                // Unregister foreground sink with the Long Running Process Singleton
+                THROW_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(processName.c_str()));
 
-                wil::unique_cotaskmem_string processName;
-                THROW_IF_FAILED(GetCurrentProcessPath(processName));
-                LOG_IF_FAILED(notificationPlatform->UnregisterLongRunningActivator(processName.get()));
-
-                s_protocolRegistration = false;
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonForegroundRegistration = false;
+                }
             }
         }
         catch (...)
         {
-            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException(), activators);
-            throw;
+            hr = wil::ResultFromCaughtException();
+            PushNotificationTelemetry::ActivatorUnregisteredByApi(hr);
```

Comment:
calling twice again

Reply:
If it fails, it will log and throw outside the try/catch. Otherwise it will log success.